crates/engine/src/ext.rs

```diff
@@ -0,0 +1,414 @@
+                storage: Storage::new(),
+                balances: HashMap::new(),
+                exec_context: ExecContext::default(),
+            });
+        );
+        ENV_INSTANCE.with(|instance| f(&mut instance.borrow_mut()))
+    }
+}
+
+macro_rules! define_error_codes {
+    (
+        $(
+            $( #[$attr:meta] )*
+            $name:ident = $discr:literal,
+        )*
+    ) => {
+        /// Every error that can be returned to a contract when it calls any of the host functions.
+        #[cfg_attr(test, derive(PartialEq, Eq))]
+        #[derive(Debug)]
+        #[repr(u32)]
+        pub enum Error {
+            $(
+                $( #[$attr] )*
+                $name = $discr,
+            )*
+            /// Returns if an unknown error was received from the host module.
+            UnknownError,
+        }
+
+        impl From<ReturnCode> for Result {
+            #[inline]
+            fn from(return_code: ReturnCode) -> Self {
+                match return_code.0 {
+                    0 => Ok(()),
+                    $(
+                        $discr => Err(Error::$name),
+                    )*
+                    _ => Err(Error::UnknownError),
+                }
+            }
+        }
+    };
+}
+define_error_codes! {
+    /// The called function trapped and has its state changes reverted.
+    /// In this case no output buffer is returned.
+    /// Can only be returned from `seal_call` and `seal_instantiate`.
+    CalleeTrapped = 1,
+    /// The called function ran to completion but decided to revert its state.
+    /// An output buffer is returned when one was supplied.
+    /// Can only be returned from `seal_call` and `seal_instantiate`.
+    CalleeReverted = 2,
+    /// The passed key does not exist in storage.
+    KeyNotFound = 3,
+    /// Transfer failed because it would have brought the sender's total balance
+    /// below the subsistence threshold.
+    BelowSubsistenceThreshold = 4,
+    /// Transfer failed for other not further specified reason. Most probably
+    /// reserved or locked balance of the sender that was preventing the transfer.
+    TransferFailed = 5,
+    /// The newly created contract is below the subsistence threshold after executing
+    /// its constructor so no usable contract instance will be created.
+    NewContractNotFunded = 6,
+    /// No code could be found at the supplied code hash.
+    CodeNotFound = 7,
+    /// The account that was called is either no contract (e.g. user account) or is a tombstone.
+    NotCallable = 8,
+}
+
+/// The raw return code returned by the host side.
+#[repr(transparent)]
+pub struct ReturnCode(u32);
+
+impl ReturnCode {
+    /// Returns the raw underlying `u32` representation.
+    pub fn into_u32(self) -> u32 {
+        self.0
+    }
+}
+
+fn extract_from_slice(output: &mut &mut [u8], new_len: usize) {
+    debug_assert!(new_len <= output.len());
+    let tmp = core::mem::take(output);
+    *output = &mut tmp[..new_len];
+}
```

Comment:
It is used by the functions (e.g. `caller`, `balance`) in this file to write their result back into `output`.

Reply:
It looks as if it is not really needed tbh since we are not looking after optimal performance but rather optimal readability in this crate. But I will take another look.